### PR TITLE
Implement tsv tools

### PR DESCRIPTION
**Pull request by 14thibea**
Tsv tools already exist in the code but  cannot be used with the general commandline.
The goal of this PR is to add the following tsv tools to the general commandline:
- **restriction**, to reproduce the restrictions applied to AIBL and OASIS datasets,
- **extract**, to extract the lists of participants for each diagnosis,
- **split**, to perform one split on extracted diagnoses (used in the original paper for train / test split),
- **kfold**, to perform a k-fold split on extracted diagnoses (used in the original paper for train / validation splits)
- **analysis**, to retrieve the demographics tables from the extracted diagnoses.

The commandline would look like this:
`clinicadl tsvtool {restriction|extract|split|kfold|analysis} <args>`

Do not hesitate to propose more appropriate names or descriptions for the tools !
I did not plan to include the `test` function which checks if there is data leakage in the splits. Please comment if you think it is needed and how it should be added.

### CODE REVIEW

**Comment by 14thibea** [comment]
I removed the intermediate folder _lists_by_diagnosis_ as this was useful when we had also a _lists_by_task_ organization but we abandonned it.

**Comment by 14thibea** [comment]
To note: the `analysis` subtask may not work well because some columns may be missing / innacurate in the output of `clinica iotools merge-tsv`. However I think it should not be modified here but in clinica.

File: README.md [hunk 1]
```diff
@@ -131,6 +136,24 @@ For detailed instructions and options of each task type  `clinica 'task' -h`.
 
 ### Some examples
 
+#### Labels extraction in tsv files
+
+Typical use for `tsvtool extract`:
+
+```text
+clinicadl tsvtool extract <merged_tsv> <missing_mods> <results_path> --restriction_path <restriction_path>
```
Comment (mdiazmel) [inline README.md]: I propose to change the `tsvtool extract` option by something like `tsvtool getlabels` in order to use a different keyword from `extract`. This keyword is already used on the hte first level options (`clinicadl extract`) and users can get easily confused.

File: clinicadl/clinicadl/cli.py [hunk 2]
```diff
@@ -695,17 +754,173 @@ def parse_command_line():
...
+        type=str)
+    tsv_restrict_subparser.add_argument(
+        "results_path",
+        help="Path to the resulting tsv file (filename included).",
+        type=str)
+
+    tsv_restrict_subparser.set_defaults(func=tsv_restrict_func)
+
+    tsv_extract_subparser = tsv_subparser.add_parser(
+        'extract',
```
Comment (mdiazmel) [inline clinicadl/clinicadl/cli.py]: _c.f._ above comment about using `extract` keyword here!

File: clinicadl/clinicadl/tools/tsv/data_formatting.py [hunk 3]
```diff
@@ -288,58 +293,29 @@ def apply_restriction(bids_df, restriction_path):
...
-                        default=['AD', 'CN'], help="Diagnosis that must be selected from the tsv file")
-    parser.add_argument("--time_horizon", default=36, type=int,
-                        help="Time horizon to analyse stability of MCI subjects.")
-    parser.add_argument("--restriction_path", type=str, default=None,
-                        help="Path to a tsv containing the sessions that can be included")
-
-    args = parser.parse_args()
-
+def tsv_extraction(merged_tsv, missing_mods, results_path,
+                   diagnoses, modality, restriction_path, time_horizon):
```
Comment (mdiazmel) [inline clinicadl/clinicadl/tools/tsv/data_formatting.py]: As we talked before, I think that is necessary to add well written docstrings to these functions (probably copy/paste some portions of the argument parsing in ` clinicadl/clinicadl/cli.py`) if possible with more details. These functions will be some kind of API for advanced users. Future documentation cab be created from the docstrings.

File: clinicadl/clinicadl/tools/tsv/data_split.py [hunk 4]
```diff
@@ -67,48 +71,20 @@ def create_split(diagnosis, diagnosis_df, merged_df, n_test,
...
-    parser.add_argument("--t_val_threshold", "-t", default=0.0642, type=float,
-                        help="The threshold used for the chi2 test.")
-    parser.add_argument("--p_val_threshold", "-p", default=0.80, type=float,
-                        help="The threshold used for the T-test.")
-    parser.add_argument("--subset_name", type=str, default="test",
-                        help="Name of the subset that is complementary to train.")
-
-    args = parser.parse_args()
-
+def split_diagnoses(merged_tsv, formatted_data_path,
```
Comment (mdiazmel) [inline clinicadl/clinicadl/tools/tsv/data_split.py:75]: Same comment that above about the docstring of the function.

File: clinicadl/clinicadl/tools/tsv/demographics_analysis.py [hunk 5]
```diff
@@ -9,51 +9,42 @@
...
 import pandas as pd
 from .tsv_utils import first_session, next_session, add_demographics
 import os
 from os import path
 import numpy as np
 
-if __name__ == "__main__":
 
-    parser = argparse.ArgumentParser(description="Argparser for data formatting")
+def demographics_analysis(merged_tsv, formatted_data_path, results_path,
```
Comment (mdiazmel) [inline clinicadl/clinicadl/tools/tsv/demographics_analysis.py:19]: Same comment, docstrings.

File: clinicadl/clinicadl/tools/tsv/kfold_split.py [hunk 6]
```diff
@@ -3,50 +3,32 @@
...
-    parser.add_argument("--n_splits", type=int, default=5,
-                        help="Define the number of subjects to put in test set."
-                             "If 0, there is no training set and the whole dataset is considered as a test set.")
-    parser.add_argument("--MCI_sub_categories", action="store_true", default=False,
-                        help="Manage MCI sub-categories to avoid data leakage")
-    parser.add_argument("--subset_name", type=str, default="validation",
-                        help="Name of the subset that is complementary to train.")
 
-    args = parser.parse_args()
+def split_diagnoses(formatted_data_path,
```
Comment (mdiazmel) [inline clinicadl/clinicadl/tools/tsv/kfold_split.py:14]: Same one, docstrings.

File: clinicadl/clinicadl/tools/tsv/demographics_analysis.py [hunk 7]
```diff
@@ -9,51 +9,42 @@
...
+
+    if age_name not in merged_tsv.columns.values:
+        raise ValueError("The column corresponding to age is not labelled as %s. "
+                         "Please change the parameters." % age_name)
+
+    if mmse_name not in merged_tsv.columns.values:
+        raise ValueError("The column corresponding to mmse is not labelled as %s. "
+                         "Please change the parameters." % mmse_name)
 
     columns = ['n_subjects', 'mean_age', 'std_age', 'min_age', 'max_age', 'sexF', 'sexM', 'mean_MMSE', 'std_MMSE',
```
Comment (mdiazmel) [inline clinicadl/clinicadl/tools/tsv/demographics_analysis.py]: Shorter lines improves readability of code, e.g.:
```suggestion
    columns = ['n_subjects', 'mean_age', 'std_age',
               'min_age', 'max_age', 'sexF',
               'sexM', 'mean_MMSE', 'std_MMSE',
               'min_MMSE', 'max_MMSE', 'CDR_0',
               'CDR_0.5', 'CDR_1', 'CDR_2',
               'CDR_3', 'mean_scans', 'std_scans',
               'n_scans']
```

File: clinicadl/clinicadl/tools/tsv/demographics_analysis.py [hunk 8]
```diff
@@ -9,51 +9,42 @@
...
                'min_MMSE', 'max_MMSE', 'CDR_0', 'CDR_0.5', 'CDR_1', 'CDR_2', 'CDR_3', 'mean_scans', 'std_scans',
```
Comment (mdiazmel) [inline clinicadl/clinicadl/tools/tsv/demographics_analysis.py]: To remove if last suggestion is accepted.
```suggestion

```

File: clinicadl/clinicadl/tools/tsv/demographics_analysis.py [hunk 9]
```diff
@@ -9,51 +9,42 @@
...
                'n_scans']
```
Comment (mdiazmel) [inline clinicadl/clinicadl/tools/tsv/demographics_analysis.py:61]: To remove if last suggestion is accepted.
```suggestion

```

File: clinicadl/clinicadl/tools/tsv/restriction.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
+# coding: utf8
+
+import pandas as pd
+from copy import deepcopy
+
+
+def aibl_restriction(merged_tsv, results_path):
```
Comment (mdiazmel) [inline clinicadl/clinicadl/tools/tsv/restriction.py:7]: Add a short paragraph to describe shortly these restrictions on AIBL.

File: clinicadl/clinicadl/tools/tsv/restriction.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    for subject, subject_df in merged_df.groupby(level=0):
+        examination_df = subject_df['examination_date']
+        dates_list = examination_df.values
+        if (dates_list == '-4').all():
+            results_df.drop(subject, inplace=True)
+
+    results_df.to_csv(results_path, sep='\t')
+
+
+def oasis_restriction(merged_tsv, results_path):
```
Comment (mdiazmel) [inline clinicadl/clinicadl/tools/tsv/restriction.py:32]: Short description of the Oasis' restrictions.

**Review by mdiazmel — COMMENTED** [review]
Hi @14thibea,

I added some comments to your PR. 

I'm wondering if removing the folder `lists_by_diagnosis` can affect the functions performing the training (`clinicadl train`). Can you confirm this (or maybe deny), please ?

Thanks,

File: clinicadl/clinicadl/cli.py (hunk 2 above)
Comment (14thibea) [reply clinicadl/clinicadl/cli.py]: Changed in commit https://github.com/aramis-lab/AD-DL/pull/18/commits/fc883c36595395546c80bd01e75b82e51f1f3764

File: README.md (hunk 1 above)
Comment (14thibea) [reply README.md]: resolved by commit https://github.com/aramis-lab/AD-DL/pull/18/commits/b9cacca8a0e622a92730bd75aec20bffee0405f9

**Comment by 14thibea** [comment]
Can you tell me if you are OK with the format of the docstring in the last commit https://github.com/aramis-lab/AD-DL/pull/18/commits/ce4435b7fd027d371bbb4207c12d36003f0b405e ? If not can you give me an example ?

**Comment by 14thibea** [comment]
The removal of `lists_by_diagnosis` only impacts the `.sh` files in `scripts`. I find it easier to use without (I always forgot it when I wrote a new launcher), however it is true that it can be useful to stay compliant with AD-ML standard, in which data is organized by diagnosis or tasks lists.

**Review by mdiazmel — COMMENTED** [review]
Thanks @14thibea ,
I'm going to solve the conflicts with master and I'll merge your PR!

**Comment by 14thibea** [comment]
Actually I am currently adding some changes to the test suite to use it more easily (without adding it to the commandline) could you maybe wait until I finish this ?

**Comment by mdiazmel** [comment]
> Actually I am currently adding some changes to the test suite to use it more easily (without adding it to the commandline) could you maybe wait until I finish this ?

Yes, no problem. Ping me here or let me know when the PR is ready to merge!

**Comment by 14thibea** [comment]
It's OK now, I added the test suite.
I also retested everything with the OASIS dataset and fixed some bugs accordingly (I tested everything with ADNI the first time).
